src/util.rs

```diff
@@ -62,3 +65,37 @@ impl LazyBool {
         self.0.unsync_init(|| init() as usize) != 0
     }
 }
+
+/// Polyfill for `maybe_uninit_slice` feature's
+/// `MaybeUninit::slice_assume_init_mut`. Every element of `slice` must have
+/// been initialized.
+#[inline(always)]
+pub unsafe fn slice_assume_init_mut<T>(slice: &mut [MaybeUninit<T>]) -> &mut [T] {
+    // SAFETY: `MaybeUninit<T>` is guaranteed to be layout-compatible with `T`.
+    mem::transmute(slice)
+}
+
+#[inline]
+pub fn uninit_slice_fill_zero(slice: &mut [MaybeUninit<u8>]) -> &mut [u8] {
+    slice.iter_mut().for_each(|b| *b = MaybeUninit::zeroed());
```

Comment:
Drive by review: this could be changed to use [`ptr::write_bytes`](https://doc.rust-lang.org/nightly/core/ptr/fn.write_bytes.html) to help when compiling in debug mode. In release mode LLVM should already be smart enough to optimize the loop into a `memset`.

Reply:
Looking at [the assembly output](https://rust.godbolt.org/z/v3nYnh458) it seems like at opt-levels `0`, `1`, and `z`, the only way to avoid this being a byte-by-byte loop is to use `ptr::write_bytes`.

It requires a line of `unsafe`, but I think that's fine.
```rust
pub fn uninit_slice_fill_zero(slice: &mut [MaybeUninit<u8>]) -> &mut [u8] {
    unsafe { write_bytes(s.as_mut_ptr(), 0, s.len()) };
    unsafe { slice_assume_init_mut(slice) }
}
```
isn't that bad.

Reply:
> unsafe { write_bytes(s.as_mut_ptr(), 0, s.len()) };

This is basically `slice::fill` as added in Rust 1.50.0; see https://github.com/rust-lang/rust/pull/70752.

Reply:
As noted in the assembly output link above, using `slice::fill` doesn't really generate different assembly than the normal `for` loop or iterator `for_each`, it's just shorter to write. That was the reason I agreed with using `ptr::write_bytes`.